### PR TITLE
Added the script for generating QAOA circuits for QUBO instances

### PR DESCRIPTION
**Pull request by Guoming-Wang**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

- Context: why is it needed? In case of bug, what was the cause for the bug?
- Concise description of the implemented solution.
- If any dependencies are added, list them and justify why they are needed.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

File: src/benchq/algorithms/_qaoa.py
Comment (max-radin) [inline src/benchq/algorithms/_qaoa.py:1]: It looks like the QAOA functions aren't part of the public API right now. We could rename `_qaoa.py` to `qaoa.py`, or import the QAOA functions in `__init__.py`.

**Review by max-radin — CHANGES_REQUESTED** [review]
Left a few minor comments.

File: src/benchq/algorithms/_qaoa.py
Comment (AthenaCaesura) [reply src/benchq/algorithms/_qaoa.py:1]: I was really just following the convention that other algorithms had at the time. Seems like this format was outdated. I'll rename it to just `qaoa.py`.

Comment (AthenaCaesura) [reply src/benchq/algorithms/_qaoa.py:1]: I was really just following the convention that other algorithms had at the time. Seems like this format was outdated. I'll rename it to just `qaoa.py`.

File: src/benchq/problem_ingestion/hamltonian_from_file.py
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/problem_ingestion/hamltonian_from_file.py:1]: Done!

File: tests/benchq/algorithms/test_qaoa.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            assert circuit.operations[i * 11 + 10].qubit_indices == (1,)
+
+            assert circuit.operations[i * 11 + 11].gate.name == "RZ"
+            assert circuit.operations[i * 11 + 11].qubit_indices == (1,)
+
+            assert circuit.operations[i * 11 + 12].gate.name == "H"
+            assert circuit.operations[i * 11 + 12].qubit_indices == (1,)
+
+
+def test_get_qaoa_circuit_produces_correct_output_for_meduim_coefficeints():
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply tests/benchq/algorithms/test_qaoa.py:66]: done!

File: src/benchq/algorithms/_qaoa.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    get_qaoa_program returns a circuit with the same randomized values for each layer.
+
+    Input:
+          hamiltonian (np.array): hamiltonian to simulate
+          n_layers (int): Number of layers in QAOA circuits
+    Output:
+          a QAOA circuit with random parameters for the QUBO instance
+    """
+
+    qaoa = QAOA.default(cost_hamiltonian=hamiltonian, n_layers=1)
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/algorithms/_qaoa.py:26]: Good catch! Would have confused a lot of people later on. :sw

File: src/benchq/problem_ingestion/hamltonian_from_file.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    else:
+        raise ValueError(
+            f"Hamiltonian extraction failed for {file_name}. "
+            f"File format is not recognized. Please use formatting from either "
+            f"Guoming Wang's QAOA implementation or Alex Kunitsa's molecule "
+            f"implementation."
+        )
+
+
+def _guoming_qaoa_hamiltonian_from_hdf5(data: h5py.File) -> PauliSum:
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/problem_ingestion/hamltonian_from_file.py:128]: Done.

File: src/benchq/problem_ingestion/hamltonian_from_file.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            pauli_terms.append(off_diag_pauli_term1)
+            pauli_terms.append(off_diag_pauli_term2)
+            pauli_terms.append(off_diag_pauli_term3)
+            pauli_terms.append(off_diag_pauli_term4)
+
+    hamiltonian = PauliSum(pauli_terms).simplify()
+    return hamiltonian
+
+
+def _alex_molecule_hamiltonian_from_hdf5(data: h5py.File) -> PauliSum:
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/problem_ingestion/hamltonian_from_file.py:158]: Done.

**Review by max-radin — APPROVED** [review]

**Pull request merged**